ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/permissions/PeerPermissionsBlacklist.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.permissions;
+
+import tech.pegasys.pantheon.ethereum.p2p.peers.Peer;
+import tech.pegasys.pantheon.util.LimitedSet;
+import tech.pegasys.pantheon.util.LimitedSet.Mode;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+
+import io.vertx.core.impl.ConcurrentHashSet;
+
+public class PeerPermissionsBlacklist extends PeerPermissions {
+  private static int DEFAULT_INITIAL_CAPACITY = 20;
+
+  private final Set<BytesValue> blacklist;
+
+  private PeerPermissionsBlacklist(final int initialCapacity, final OptionalInt maxSize) {
+    if (maxSize.isPresent()) {
+      blacklist =
+          LimitedSet.create(initialCapacity, maxSize.getAsInt(), Mode.DROP_LEAST_RECENTLY_ACCESSED);
```

Comment:
Consider Guava's Cache instead: https://github.com/google/guava/wiki/CachesExplained - it doesn't do Least Recently Accessed 100% but it does have better concurrency then being wrapped in a synchronized set.

Under the covers Java's sets are Maps with the key being the entry and I can't remember what the value is.

Reply:
I may follow-up on this suggestion in another PR.  But skipping for now.